### DISCUSSION TITLE
How to entend rules in forwardRef?

### DISCUSSION THREAD

**Discussion by geoguide** [react-hook-form/react-hook-form#2412] [Q&A] [answered]
I have custom components that I'm using forwardRef on from the form on a page. I want to be able to pass rules from both the page and the component.

Component:
```javascript
<InputMask
        placeholder={'YYYY/MM/DD'}
        mask='9999/99/99'
        name={name}
        className={classnames(inputClassNames)}
        ref={ref}
        type={type}
        disabled={disabled}
        rules={{
          pattern: /([12]\d{3}\/(0[1-9]|1[0-2])\/(0[1-9]|[12]\d|3[01]))/i,
          ...rules,
        }}
/>
```

Usage on page:
```javascript
<DateInput
        name='triggeredAt'
        label='Date'
        touched={touched}
        errors={errors}
        ref={register}
        required={{ required: true }}
/>
```

So I always want to validate the pattern to make sure it's a date, but not all versions of this input will be required. How can I achieve this?

I am aware I could pass a boolean to the component and build the rules there (edit: checking again this does not seem possible) but what if I wanted to pass arbitrary rules? Thanks.

**Comment by bluebill1049** [comment]
use validate function.

**Reply by geoguide** [reply]
Wouldn't I still need to define the validate function in the page?

**Comment by geoguide** [comment, marked as answer]
The solution is to not pass ref, make the subcomponent a functional component and pass it register and rules.

```
<InputMask
  placeholder={'YYYY/MM/DD'}
  mask='9999/99/99'
  name={name}
  className={classnames(inputClassNames)}
  register={register}
  type={type}
  disabled={disabled}
  rules={{
    pattern: /([12]\d{3}\/(0[1-9]|1[0-2])\/(0[1-9]|[12]\d|3[01]))/i,
    ...rules,
  }}
/>

...

<DateInput
  name='triggeredAt'
  label='Date'
  touched={touched}
  errors={errors}
  register={register}
  rules={{ required: true }}
/>
```

**Reply by bluebill1049** [reply]
nice.
